packages/cli/src/commands/devHandler.js

```diff
@@ -23,6 +45,97 @@ export const handler = async ({
 }) => {
   const rwjsPaths = getPaths()
 
+  let apiPort = getConfig().api.port
+  let webPort = getConfig().web.port
+
+  // Check api port
+  if (side.includes('api')) {
+    const freePort = await getFreePort(apiPort)
+    if (freePort === -1) {
+      console.error(
+        c.error(
+          `Requested API port of ${apiPort} is already in use and no neighbouring port is available! Cannot start development server.`
+        )
+      )
+      process.exit(1)
+    }
+    if (freePort !== apiPort) {
+      console.log(
+        c.warning(
+          `Requested API port of ${apiPort} is already in use however ${freePort} is available.`
+        )
+      )
+      const useAvailablePort = await prompts({
+        type: 'confirm',
+        name: 'port',
+        message: `Do you wish to use port ${freePort} instead?`,
+        initial: true,
+        active: 'Yes',
+        inactive: 'No',
+      })
+      if (!useAvailablePort.port) {
+        console.log(c.info(`The API port can be updated in 'redwood.toml'.`))
+        process.exit(0)
+      }
+    }
+    apiPort = freePort
+
+    // TODO: Check the apiDebugPort too?
+  }
+
+  // Check web port
+  if (side.includes('web')) {
+    // Check for specific forwarded web port
+
+    const forwardedPortMatches = forward.match(
+      /--port=[0-9][0-9]?[0-9]?[0-9]?[0-9]? ?/
+    )
+    const forwardedPortSet =
+      forwardedPortMatches && forwardedPortMatches.length == 1
+    if (forwardedPortSet) {
+      webPort = forwardedPortMatches[0]
+        .substring(forwardedPortMatches[0].indexOf('=') + 1)
+        .trim()
+    }
```

Comment:
I thought about a better way to do this. Instead of patching on the value of port, let's just match on the flag and take the value the user specified. Also I tested what webpack server does if `--port` is passed twice—it uses the last one specified. So I'm doing a `matchAll` with a global regex, and then `pop` will give us the last one (provided there's a match):

```suggestion
    const forwardedPortMatches = [
      ...forward.matchAll(/\-\-port(\=|\s)(?<port>[^\s]*)/g),
    ]

    const forwardedPortSet = Boolean(forwardedPortMatches.length)

    if (forwardedPortSet) {
      webPort = forwardedPortMatches.pop().groups.port
    }
```

Reply:
I trust your regex magic 😆. I added your suggested change and tested it out, worked great.